planning/mission_planner/src/lanelet2_plugins/default_planner.cpp

```diff
@@ -186,21 +194,114 @@ PlannerPlugin::MarkerArray DefaultPlanner::visualize(const HADMapRoute & route)
   return route_marker_array;
 }
 
-bool DefaultPlanner::is_goal_valid(const geometry_msgs::msg::Pose & goal) const
+visualization_msgs::msg::MarkerArray DefaultPlanner::visualize_debug_footprint(
+  tier4_autoware_utils::LinearRing2d goal_footprint) const
+{
+  visualization_msgs::msg::MarkerArray msg;
+  auto marker = tier4_autoware_utils::createDefaultMarker(
+    "map", rclcpp::Clock().now(), "goal_footprint", 0, visualization_msgs::msg::Marker::LINE_STRIP,
+    tier4_autoware_utils::createMarkerScale(0.05, 0.0, 0.0),
+    tier4_autoware_utils::createMarkerColor(0.99, 0.99, 0.2, 1.0));
+  marker.lifetime = rclcpp::Duration::from_seconds(2.5);
+
+  marker.points.push_back(
+    tier4_autoware_utils::createPoint(goal_footprint[0][0], goal_footprint[0][1], 0.0));
+  marker.points.push_back(
+    tier4_autoware_utils::createPoint(goal_footprint[1][0], goal_footprint[1][1], 0.0));
+  marker.points.push_back(
+    tier4_autoware_utils::createPoint(goal_footprint[2][0], goal_footprint[2][1], 0.0));
+  marker.points.push_back(
+    tier4_autoware_utils::createPoint(goal_footprint[3][0], goal_footprint[3][1], 0.0));
+  marker.points.push_back(marker.points.front());
+
+  msg.markers.push_back(marker);
+
+  return msg;
+}
+
+bool DefaultPlanner::check_goal_footprint(
+  const lanelet::ConstLanelet & current_lanelet,
+  const lanelet::ConstLanelet & combined_prev_lanelet,
+  const tier4_autoware_utils::Polygon2d & goal_footprint, double & next_lane_length,
+  const double search_margin)
+{
+  std::vector<tier4_autoware_utils::Point2d> points_intersection;
+
+  // check if goal footprint is in current lane
+  boost::geometry::intersection(
+    goal_footprint, combined_prev_lanelet.polygon2d().basicPolygon(), points_intersection);
+  if (points_intersection.empty()) {
+    return true;
+  }
+  points_intersection.clear();
+
+  // check if goal footprint is in between many lanelets
+  for (const auto & next_lane : routing_graph_ptr_->following(current_lanelet)) {
+    next_lane_length += lanelet::utils::getLaneletLength2d(next_lane);
+    lanelet::ConstLanelets lanelets;
+    lanelets.push_back(combined_prev_lanelet);
+    lanelets.push_back(next_lane);
+    lanelet::ConstLanelet combined_lanelets = combine_lanelets(lanelets);
+
+    // if next lanelet length longer than vehicle longitudinal offset
+    if (vehicle_info_.max_longitudinal_offset_m + search_margin < next_lane_length) {
+      next_lane_length -= lanelet::utils::getLaneletLength2d(next_lane);
+      boost::geometry::intersection(
+        goal_footprint, combined_lanelets.polygon2d().basicPolygon(), points_intersection);
+      if (points_intersection.empty()) {
+        return true;
+      }
+      points_intersection.clear();
+    } else {  // if next lanelet length shorter than vehicle longitudinal offset -> recursive call
+      if (!check_goal_footprint(next_lane, combined_lanelets, goal_footprint, next_lane_length)) {
```

Comment:
I feel that recursive call of a bit compress if/else condition is hard to debug, so I do not want to use it if we can.

In `check_goal_footprint`, how about implementing like
- 1. get goal_lanelet, and hold it as target_lanelets
- 2. add next lanelets from the goal to target_lanelets until the additional lane length is longer than vehicle_info_.max_longitudinal_offset_m + magin
- 3. add prev lanelets from the goal to target_lanelets until the additional lane length is longer than vehicle_info_.**min**_longitudinal_offset_m + magin
- 4. check if the goal footprint intersects with target_lanelets.

Reply:
First of all, many thanks for the review, I think it is necessary to write a recursive function in this point. Because when lanelets branches as in the case below, I think we need to use a recursive structure to search on all these cases.

Yes, you are right that the debugging part is difficult. To debug this function, I checked many things, such as the lanelets id's traversed on the edge cases, the length of the total traversed branches, etc. So in this final version I think everything is working as it should.

For such a search method, using path_lanelets as prev lanelets makes the algorithm simpler. If we also search for each prev_lanelets, we'll have to do a recursive search for each prev lanelet going to next lanelets

What do you think about this cases?

![check_goal_footprint drawio](https://user-images.githubusercontent.com/56237550/204151167-ab81280f-54de-4d3f-98fc-12d7f8f9dc69.png)

Reply:
I understand the reason to use recursive call. Thank you very much for the good explanation.

---

Let me summarize the current collision checking (= if the path footprints are inside the lanelets or not) in autoware.universe.
With the current autoware.universe, the lanelets after the goal_lanelet are not considered for collision checking in 
- 1. mission_planner
- 2. obstacle_avoidance_planner
    - obstacle_avoidance_planner gets a drivable area from behavior_path_planner, and check if path footprints are inside the drivable area.
    - Drivable area is generated in behavior_path_planner using lanelets from the start to goal lanelet calculated in the mission_planner

With this PR, the mission planner will consider the lanelets after the goal for collision checking. So 1 will be fixed.

However, the behavior_path_planner, which generates the drivable area, does not know the correct lanelets after the goal in the case of your figure. Only mission_planner has the information of these lanelets.
Therefore, the drivable area can only have lanelets until the goal_lanelet, and the lanelets after the goal cannot be added to the drivable area. Therefore, 2 still exists.

IMO, in order to fix 2 (obstacle_avoidance_planner) as well as 1 (mission_planner), the goal lanelets should be the last lanelet in `check_goal_footprint` (maybe `next_lane` when returning true).
With this change, the drivable area can cover until the last lanelet after the goal considered in mission_planner, and 2 will be fixed.

In short, my suggestion is
- calculate the last lanelet recursively and hold it as `goal_lanelet`
    - This  implementation will be almost the same as `check_goal_footprint` in this PR.
- then, plan route between start lanelet to `goal_lanelet`.
- then, check if the goal footprint is inside the lanelets until `goal_lanelet`.


---

However, this change will be a bit large. I'm okay to merge the current PR without the further change to fix 1, and after that we will apply the change I suggested to fix 2.
Sorry to be too long. Feel free to share your opinion.

Reply:
(We can have a call if it's better for you.)

Reply:
First of all, sorry for the late conversion.

Thank you very much for your nice explanation, I understood the situation you said and we need to update it as well.

We can merge this PR if it's okay with you. I want to fix the 2nd situation by creating a new PR.

The methods you suggest can be done for this. 

- Additionally, I can create a new route by adding the traversed path up to the lane the goal is in to the path_lanelets.

- I was checking this path while debugging, it may be comfortable for me to implement it.

- As you said, I can create a new route between the last point where the goal lanelet is located and the point where the vehicle is located, that is also possible.

So how about I open a new PR with these? Or should I continue on this PR?

Reply:
For me, it's okay to implement it on a new PR.
Then it seems that this PR is ready to merge?

Reply:
Due to my late review, there happened a conflict.
I'm sorry, but could you resolve it. Then I will approve it.

Reply:
@takayuki5168 I solved conflicts, I thin it is ready to merge :+1: